core/src/main/java/org/radargun/stages/test/Stressor.java

```diff
@@ -10,112 +11,102 @@
-         try {
-            if (useTransactions && txRemainingOperations <= 0) {
-               ongoingTx = stage.transactional.getTransaction();
-               logic.transactionStarted();
+         while (!test.isFinished()) {
+            // one cycle of this loop should match to a stage.
+
+            // this is the ramp-up state
+            while (!test.isFinished() && !test.isSteadyState()) {
+               selector = test.getSelector();
+               try {
+                  Conversation conversation = selector.next();
+                  conversation.run(this);
+               } catch (InterruptedException e) {
+                  // just react on test state
+               } catch (RequestException e) {
+                  // the exception was already logged in makeRequest
+               }
             }
-            result = logic.run();
-         } catch (OperationLogic.RequestException e) {
-            // the exception was already logged in makeRequest
-         }
-         i++;
-         completion.logProgress(i);
-         if (i % stage.logPeriod == 0) TestStage.avoidJit(result);
-      }
 
-      if (txRemainingOperations > 0) {
-         try {
-            long endTxTime = endTransaction();
-            stats.registerRequest(endTxTime, stage.commitTransactions ? Transactional.COMMIT : Transactional.ROLLBACK);
-            stats.registerRequest(transactionDuration + endTxTime, Transactional.DURATION);
-         } catch (TransactionException e) {
-            stats.registerError(e.getOperationDuration(), stage.commitTransactions ? Transactional.COMMIT : Transactional.ROLLBACK);
-            stats.registerError(transactionDuration + e.getOperationDuration(), Transactional.DURATION);
-         } finally {
-            clearTransaction();
+            stats = test.createStatistics();
+            if (stats == null) {
+               // cannot create statistics because the test did not started at all
+               break;
+            }
+            stats.begin();
+            try {
+               // and this is the steady state
+               while (test.isSteadyState()) {
+                  if (selector == null) {
+                     // this means that the thread was not created during ramp-up state,
+                     // since it did not grab the selector
+                     selector = test.getSelector();
+                  }
+                  try {
+                     Conversation conversation = selector.next();
+                     conversation.run(this);
+                  } catch (InterruptedException e) {
+                     // the test is interrupted
+                  } catch (RequestException e) {
+                     // the exception was already logged in makeRequest
+                  }
+               }
+            } finally {
+               stats.end();
+               test.recordStatistics(stats);
+               stats = null; // let's throw NPE if we try to record now
```

Comment:
So is the comment still valid? Causing intentional NPE anywhere is not acceptable.